tracing-flame/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+//! the execution of your program. This representation is best created with a
+//! `flamechart`, which _does not_ sort or collapse identical stack frames.
+//!
+//! [`tracing`]: https://docs.rs/tracing
+//! [`inferno`]: https://docs.rs/inferno
+//! [`FlameLayer`]: struct.FlameLayer.html
+//! [`FlushGuard`]: struct.FlushGuard.html
+#![warn(
+    missing_debug_implementations,
+    missing_docs,
+    rust_2018_idioms,
+    unreachable_pub,
+    bad_style,
+    const_err,
+    dead_code,
+    improper_ctypes,
+    non_shorthand_field_patterns,
+    no_mangle_generic_items,
+    overflowing_literals,
+    path_statements,
+    patterns_in_fns_without_body,
+    private_in_public,
+    unconditional_recursion,
+    unused,
+    unused_allocation,
+    unused_comparisons,
+    unused_parens,
+    while_true
+)]
+
+pub use error::Error;
+
+use error::Kind;
+use std::cell::Cell;
+use std::fmt;
+use std::fmt::Write as _;
+use std::fs::File;
+use std::io::BufWriter;
+use std::io::Write;
+use std::marker::PhantomData;
+use std::path::Path;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::time::{Duration, Instant};
+use tracing::span::Attributes;
+use tracing::Id;
+use tracing::Subscriber;
+use tracing_subscriber::layer::Context;
+use tracing_subscriber::registry::LookupSpan;
+use tracing_subscriber::registry::SpanRef;
+use tracing_subscriber::Layer;
+
+mod error;
+
+thread_local! {
+    static START: Instant = Instant::now();
+    static LAST_EVENT: Cell<Instant> =  Cell::new(START.with(|f| *f));
+}
+
+/// A Layer that logs span open/close events as folded flame graph stack
+/// samples.
+///
+/// The output of `FlameLayer` emulates the output of commands like `perf` once
+/// they've been collapsed by `inferno-flamegraph`. The output of this layer
+/// should look similar to the output of the following commands:
+///
+/// ```sh
+/// perf record --call-graph dwarf target/release/mybin
+/// perf script | inferno-collapse-perf > stacks.folded
+/// ```
+///
+/// # Sample Counts
+///
+/// Because `tracing-flame` doesn't use sampling the number at the end of each
+/// folded stack trace does not represent a number of samples of that stack,
+/// instead the numbers on each line are the number of nanoseconds since the
+/// last event in the same thread.
+///
+/// # Droping and Flushing
+///
+/// Because of the way that `tracing` works, if you use a global subscriber the
+/// drop implementations on your various layers will not get called when your
+/// program exits. This means that if you're using a buffered writer as the
+/// inner writer for the `FlameLayer` you're not guaranteed to see all the
+/// events that have been emitted in the file by default.
+///
+/// To help with this `FlameLayer` exposes the [`flush_on_drop`] function, which
+/// returns a [`FlushGuard`]. The `FlushGuard` will flush on drop and if
+/// necessary can be used to manually flush the writer within the `FlameLayer`.
```

Comment:
```suggestion
/// To ensure all data is flushed when the program exits, `FlameLayer` exposes
/// the [`flush_on_drop`] function, which returns a [`FlushGuard`]. The `FlushGuard`
/// will flush the writer when it is dropped. If necessary, it can also be used to manually
/// flush the writer.
```